src/osp/coordinates.h

```diff
@@ -0,0 +1,144 @@
+/**
+ * Open Space Program
+ * Copyright © 2019-2020 Open Space Program Project
+ *
+ * MIT License
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+#pragma once
+
+#include "universetypes.h"
+#include "types.h"
+
+#include <entt/core/family.hpp>
+#include <entt/core/any.hpp>
+
+#include <Corrade/Containers/StridedArrayView.h>
+
+#include <optional>
+#include <variant>
+#include <vector>
+
+namespace osp::universe
+{
+
+using ccomp_family_t = entt::family<struct ccomp_type>;
+using ccomp_id_t = ccomp_family_t::family_type;
+
+// Coordinate component types (CComp). ccomp_family_t generates IDs for these
+// at runtime, used as indices for CoordinateSpace::m_components
+struct CCompX { using datatype_t = spaceint_t; };
+struct CCompY { using datatype_t = spaceint_t; };
+struct CCompZ { using datatype_t = spaceint_t; };
+struct CCompSat { using datatype_t = Satellite; };
+
+template <typename CCOMP_T>
+using ViewCComp_t = Corrade::Containers::StridedArrayView1D<typename CCOMP_T::datatype_t>;
+
+template<typename CCOMP_T>
+constexpr ccomp_id_t ccomp_id() noexcept
+{
+    return ccomp_family_t::type<CCOMP_T>;
+}
+
+using CoordinateView_t = std::optional<Corrade::Containers::StridedArrayView1D<void>>;
+
+/**
+ * @brief Stores positions, velocities, and other related data for Satellites,
+ *        and exposes them through read-only strided array views.
+ *
+ * CoordinateSpaces must be managed by some external system. They can store any
+ * kind of coordinate data, such as Cartesian XYZ or orbital parameters.
+ *
+ * As part of ECS, this is a way to store common components in separate buffers.
+ * that can each be managed by a specific system (simultaneously too).
+ * aka: a non-EnTT but still ECS component pool
+ *
+ */
+struct CoordinateSpace
+{
+    using SatToAdd_t = std::tuple<Satellite, Vector3g, Vector3>;
```

Comment:
Std::tuple does have a pretty hefty compilation-time cost

not enough to justify changing this line, but I've found myself much more frequently using simple POD structs now that C++ has relatively convienient support for that.

You can even continue using structured bindings with the POD type if you define a specialization of get<> for it. https://devblogs.microsoft.com/oldnewthing/20201015-00/?p=104369